### PR TITLE
Staticman nested comments support from Huginn

### PR DESCRIPTION
**Pull request by VincentTam**
## Description

It is in turn a port from halogenica/beautifulhugo#222.  (typo edited)  Introduced true nested comments with reply functionality.

1. Reply target anchors to easily jump between replies.
2. Interactive reply target display with Gravatar and name.
3. Improved input type for "website" field.
4. Implemented FR #63 to construct POST URL in JS.
5. Clearer instructions for Google reCAPTCHA v2.
6. Introduced some SCSS variables for layout control
7. Fixed missing "id" attribute in each comment.
8. Fixed missing translation UI text for Staticman forms.
9. Changed default API endpoint to the one serving @staticmanlab due to
eduardoboucas/staticman#307.

## Motivation and Context

Resolving #63 and adding missng i18n strings to Staticman form requires changing the template, so I adapted my Staticman templates for [Huginn](https://gitlab.com/VincentTam/huginn), [Introduction](//github.com/victoriadrake/hugo-theme-introduction/pull/119), [Hugo Swift Theme](https://github.com/onweru/hugo-swift-theme/) and [my blog](https://lstu.fr/3s) into this theme.

## Screenshots (if appropriate):

Screenshot from 2019-08-16 23-59-41

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the [Contributing Document](https://github.com/pacollins/hugo-future-imperfect-slim/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the [code style](https://github.com/pacollins/hugo-future-imperfect-slim/blob/master/.github/CONTRIBUTING.md#Style-Guide) of this project.
- [ ] My change requires a change to the [documentation](https://github.com/pacollins/hugo-future-imperfect-slim/wiki).
- [ ] I have updated the documentation, including `theme.toml`, accordingly.

### CODE REVIEW

File: layouts/post/staticman.html [hunk 1]
```diff
@@ -1,51 +1,119 @@
...
+    {{ if not $comments }}
+      <p>{{ i18n "no_comment" }}</p>
+    {{ else }}
+      {{/* Loop through "main comments" only if "data/comments exists" */}}
+      {{ range $comments }}
+        {{ if not .replyThread }}
+          {{ $.Scratch.Set "threadID" ._id }}
+          {{/* Comment template for a "main comment" */}}
+          <article id="{{ ._id }}" class="post-comment">
+            <img class="post-comment-avatar" src="https://www.gravatar.com/avatar/{{ .email }}?s=100">
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: Done in 41e7c689ad1cca38fa19e6ed28983015350d509e.

File: layouts/post/staticman.html [hunk 2]
```diff
@@ -1,51 +1,119 @@
...
+              <span class='post-comment-threadID hidden'>{{ ._id }}</span>
+              <a class='post-comment-reply-btn button' href='#post-js-form'>{{ i18n "reply" }}</a>
+            </div>
+          </article>
+          {{/* Loop through "comment replies" */}}
+          {{ range $comments }}
+            {{ if eq .replyThread ($.Scratch.Get "threadID") }}
+            {{/* Comment template for a "comment reply" */}}
+            <article id="{{ ._id }}" class="post-comment post-comment-reply">
+              <img class="post-comment-avatar" src="https://www.gravatar.com/avatar/{{ .email }}?s=100">
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

File: layouts/post/staticman.html [hunk 3]
```diff
@@ -1,51 +1,119 @@
...
+      <input type="hidden" name="options[entryId]" value="{{ $entryId }}">
+      <input type='hidden' name='fields[replyThread]' value=''>
+      <input type='hidden' name='fields[replyID]' value=''>
+      <input type='hidden' name='fields[replyName]' value=''>
 
-{{ $comments := readDir "data/comments" }}
-{{ $.Scratch.Add "hasComments" 0 }}
-{{ $entryId := .File.UniqueID }}
+      {{/* User input fields */}}
+      <input required name='fields[name]' type='text' placeholder='{{ i18n "name" }}'>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: N/A: hidden fields are invisible

File: layouts/post/staticman.html [hunk 4]
```diff
@@ -1,51 +1,119 @@
...
+    <form id="post-js-form" class="post-new-comment" method="POST">
+      {{/* Display reply target */}}
+      <h5 class='post-reply-notice hidden'>
+        <span class='post-reply-arrow'></span><span class='post-reply-name'></span>
+      </h5>
 
-<h2>Comments</h2>
+      {{/* Hidden fields */}}
+      <input type="hidden" name="options[entryId]" value="{{ $entryId }}">
+      <input type='hidden' name='fields[replyThread]' value=''>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: N/A: hidden fields are invisible

File: layouts/post/staticman.html [hunk 5]
```diff
@@ -1,51 +1,119 @@
...
+      <input type='hidden' name='fields[replyID]' value=''>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

File: layouts/post/staticman.html [hunk 6]
```diff
@@ -1,51 +1,119 @@
...
+      <input type='hidden' name='fields[replyName]' value=''>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

File: layouts/post/staticman.html [hunk 7]
```diff
@@ -1,51 +1,119 @@
...
 
-{{ $comments := readDir "data/comments" }}
-{{ $.Scratch.Add "hasComments" 0 }}
-{{ $entryId := .File.UniqueID }}
+      {{/* User input fields */}}
+      <input required name='fields[name]' type='text' placeholder='{{ i18n "name" }}'>
+      <input name='fields[website]' type='text' placeholder='{{ i18n "website" }}'>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: We have placeholder text here.

File: layouts/post/staticman.html [hunk 8]
```diff
@@ -1,51 +1,119 @@
...
+      <input required name='fields[email]' type='email' placeholder='{{ i18n "email" }}'>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

File: layouts/post/staticman.html [hunk 9]
```diff
@@ -1,51 +1,119 @@
...
-        {{ .body | markdownify }}
-      </div>
-    </article>
-    {{ end }}
-{{ end }}
-{{ end }}
+      {{/* reCAPTHCA v2 */}}
+      {{ if and .Site.Params.staticman.recaptcha.siteKey .Site.Params.staticman.recaptcha.encryptedKey }}
+      {{ with .Site.Params.staticman.recaptcha }}
+        <input hidden name="options[reCaptcha][siteKey]" value="{{ .sitekey }}">
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

File: layouts/post/staticman.html [hunk 10]
```diff
@@ -1,51 +1,119 @@
...
+      <input type='hidden' name='fields[replyName]' value=''>
 
-{{ $comments := readDir "data/comments" }}
-{{ $.Scratch.Add "hasComments" 0 }}
-{{ $entryId := .File.UniqueID }}
+      {{/* User input fields */}}
+      <input required name='fields[name]' type='text' placeholder='{{ i18n "name" }}'>
+      <input name='fields[website]' type='text' placeholder='{{ i18n "website" }}'>
+      <input required name='fields[email]' type='email' placeholder='{{ i18n "email" }}'>
+      <textarea required name='fields[body]' placeholder='{{ i18n "body" }}' rows='10'></textarea>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

File: layouts/post/staticman.html [hunk 11]
```diff
@@ -1,51 +1,119 @@
...
+      <p class='post-submit-notice'>
+        <strong class='post-submit-notice-text submit-success hidden'>{{ i18n "success_msg" }}</strong>
+        <strong class='post-submit-notice-text submit-failed hidden'>{{ i18n "error_msg" }}</strong>
+      </p>
 
-{{ if eq ($.Scratch.Get "hasComments") 0 }}
-  <p>Nothing yet.</p>
-{{ end }}
+      {{/* Form buttons */}}
+      <input type='submit' value='{{ i18n "btn_sumbit" }}' class='button'>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: N/A: submit button

File: layouts/post/staticman.html [hunk 12]
```diff
@@ -1,51 +1,119 @@
...
+      <input type='submit' value='{{ i18n "btn_submitted" }}' class='hidden button' disabled>
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: N/A: submit button

File: layouts/post/staticman.html [hunk 13]
```diff
@@ -1,51 +1,119 @@
...
+  <h2>{{ i18n "say_sth" }}</h2>
+    <form id="post-js-form" class="post-new-comment" method="POST">
+      {{/* Display reply target */}}
+      <h5 class='post-reply-notice hidden'>
+        <span class='post-reply-arrow'></span><span class='post-reply-name'></span>
+      </h5>
 
-<h2>Comments</h2>
+      {{/* Hidden fields */}}
+      <input type="hidden" name="options[entryId]" value="{{ $entryId }}">
```
[the remark this answers is not in the document]
Comment (VincentTam) [reply layouts/post/staticman.html]: idem

**Comment by pacollins** [comment]
Marginally related: I have removed the Accesslint App for the time being. I may reintroduce it once things settle and we can focus on accessibility.

**Comment by VincentTam** [comment]
@pacollins Please review my PR.  Here's some (possible) bugs in the current Staticman template.

https://github.com/pacollins/hugo-future-imperfect-slim/blob/afbf3ccbc8889f3dbacbade26c8bf4bbb18d2e64/layouts/post/staticman.html#L2-L4'

Even though this PR is not as important as #47 and #71, the possible spamming problem mentioned in #63 can be quite annoying.  For example, you may see https://github.com/hashtafak/hashtafak.github.io/commits/master.  I confess that I didn't know Node.JS when I created my Staticman instance for testing.  I've tried ways like reCAPTCHA v2 and `allowedOrigins`, but these aren't really effective.  Note that Akismet for official instances has probably been disabled: https://github.com/eduardoboucas/staticman/issues/83#issuecomment-284173224.  Moving the form action URL to JS is the next way that I've found from https://discourse.gohugo.io/t/disqus-alternatives/2948/42?u=vincenttam.

https://github.com/pacollins/hugo-future-imperfect-slim/blob/afbf3ccbc8889f3dbacbade26c8bf4bbb18d2e64/layouts/post/staticman.html#L23

Use `.Site.Data.comments` instead.  This would allow us to remove `data/comments/.gitkeep`.  (I didn't realise that you put `readDir` when I proposed to remove `data/comments/.gitkeep` in one of my former PRs.)  Not every user want to have comments.

https://github.com/pacollins/hugo-future-imperfect-slim/blob/afbf3ccbc8889f3dbacbade26c8bf4bbb18d2e64/layouts/post/staticman.html#L31

`{{ .uniqueID }}` undefined.

https://github.com/pacollins/hugo-future-imperfect-slim/blob/afbf3ccbc8889f3dbacbade26c8bf4bbb18d2e64/layouts/post/staticman.html#L33-L35

`website` isn't a required field.  In case of empty `website`, a link to the current post head is created.  That's not something desired.

**Comment by pacollins** [comment]
I will review now. I needed time to properly test (which I have at this moment).

**Comment by pacollins** [comment]
As of right now, using @staticmanlab:
1. API set to `staticman3.herokuapp.com`
2. Bot set as collaborator
3. Bot linked via `https://staticman3.herokuapp.com/v2/connect/`
4. Received `OK`
5. Submit comment and receive `Sorry, there was an error with your submission. Please make sure all required fields have been completed and try again.`

Form HTML is from the localhost:
```
<form id="post-js-form" class="post-new-comment" method="POST" _lpchecked="1">
      
      <h5 class="post-reply-notice hidden">
        <span class="post-reply-arrow"></span><span class="post-reply-name"></span>
      </h5>

      
      <input type="hidden" name="options[entryId]" value="6354f801456983b22c4b91ce42adfca7">
      <input type="hidden" name="fields[replyThread]" value="">
      <input type="hidden" name="fields[replyID]" value="">
      <input type="hidden" name="fields[replyName]" value="">

      
      <input required="" name="fields[name]" type="text" placeholder="Your name (Required)" style="background-image: url(&quot;data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABHklEQVQ4EaVTO26DQBD1ohQWaS2lg9JybZ+AK7hNwx2oIoVf4UPQ0Lj1FdKktevIpel8AKNUkDcWMxpgSaIEaTVv3sx7uztiTdu2s/98DywOw3Dued4Who/M2aIx5lZV1aEsy0+qiwHELyi+Ytl0PQ69SxAxkWIA4RMRTdNsKE59juMcuZd6xIAFeZ6fGCdJ8kY4y7KAuTRNGd7jyEBXsdOPE3a0QGPsniOnnYMO67LgSQN9T41F2QGrQRRFCwyzoIF2qyBuKKbcOgPXdVeY9rMWgNsjf9ccYesJhk3f5dYT1HX9gR0LLQR30TnjkUEcx2uIuS4RnI+aj6sJR0AM8AaumPaM/rRehyWhXqbFAA9kh3/8/NvHxAYGAsZ/il8IalkCLBfNVAAAAABJRU5ErkJggg==&quot;); background-repeat: no-repeat; background-attachment: scroll; background-size: 16px 18px; background-position: 98% 50%; cursor: auto;">
      <input name="fields[website]" type="text" placeholder="Your website">
      <input required="" name="fields[email]" type="email" placeholder="Your email address (Required for Gravatar)">
      <textarea required="" name="fields[body]" placeholder="Your message. Feel free to use Markdown (Google 'Markdown Cheat Sheet')." rows="10"></textarea>

      
      

      
      <p class="post-submit-notice">
        <strong class="post-submit-notice-text submit-success hidden">Thanks for your comment! It will show on the site once it has been approved.</strong>
        <strong class="post-submit-notice-text submit-failed">Sorry, there was an error with your submission.  Please make sure all required fields have been completed and try again.</strong>
      </p>

      
      <input type="submit" value="Submit" class="button">
      <input type="submit" value="Submitted" class="button hidden" disabled="">
      <input type="reset" value="Reset" class="button">
    </form>
```

**Comment by VincentTam** [comment]
Perhaps some images like

comment reply diagram

might help understanding my code.

To give a *rough idea* about the entire script, there're three parts in the Staticman script.

1. AJAX post request: https://github.com/eduardoboucas/popcorn/blob/gh-pages/js/main.js
    1. Form submit event is detected
    2. Submit button is disabled
    3. Construct form action URL from several JS variables (to avoid search bots from obtaining the action URL by scanning HTML files)
    2. Form data converted to serialized URL by jQuery
    3. Submit form data
    4. Handle resutls
        - success: display success message for 3 sec + clear form
        - failed: display error message
2. Nested comments: https://vincenttam.gitlab.io/post/2018-11-17-nested-comments-in-beautiful-hugo/

    1.  In the Go-HTML template, use an `<article>` tag whose `id` attribute is set to `_id` of the YML/JSON data file to represent each comment.
    2. Use event delegation for setting the reply target: register the `<article>` tag, which listens to clicks on the reply button in each comment.  The event delegate target is the `<article>` tag.  I've used a trick of storing the thread ID (the `_id` of the eldest ancestor) in the reply buttons (which is an `<a>` internal reference in the `title` attribute).  When a reply button is clicked, the browser script searches for attributes/text in the 
    3. 

    ```js
    // record reply target when "reply to this comment" is pressed
    $('article.static-comment').on('click', 'a.reply-btn', function (evt){
      var cmt = $(evt.delegateTarget);
      $('input[name="fields[replyThread]"]').val(this.title);
      $('input[name="fields[replyID]"]').val(cmt.attr("id"));
      authorTag = cmt.find('.comment-author');
      replyName = authorTag.text();
      $('input[name="fields[replyName]"]').val(replyName);
      $('.reply-notice-text').text('↷\xa0' + replyName);
    });
   ```
    :man_technologist: https://gitlab.com/VincentTam/huginn/blob/c9e48180845fc19f8c111d551ba8b3faa67aa927/assets/js/staticman.js#L47-56

**Comment by pacollins** [comment]
Not gonna lie, Staticman goes over my ability. I still can't get a comment to post. Receiving the following from the console:

```
POST https://staticman3.herokuapp.com/v3/entry/github/pacollins/imperfect-test/master/comments 500 (Internal Server Error)
```
and
```
responseJSON:
errorCode: "GITHUB_READING_FILE"
rawError:
message: "{"message":"Not Found","documentation_url":"https://developer.github.com/v3/repos/contents/#get-contents"}"
statusCode: 404
__proto__: Object
success: false
__proto__: Object
responseText: "{"success":false,"rawError":{"message":"{\"message\":\"Not Found\",\"documentation_url\":\"https://developer.github.com/v3/repos/contents/#get-contents\"}","statusCode":404},"errorCode":"GITHUB_READING_FILE"}"
```

**Comment by VincentTam** [comment]
> As of right now, using @staticmanlab:
> 
> 1. API set to `staticman3.herokuapp.com`
> 2. Bot set as collaborator
> 3. Bot linked via `https://staticman3.herokuapp.com/v2/connect/`
> 4. Received `OK`
> 5. Submit comment and receive `Sorry, there was an error with your submission. Please make sure all required fields have been completed and try again.`
> 
> Form HTML is from the localhost:
> 
> ```
> <form id="post-js-form" class="post-new-comment" method="POST" _lpchecked="1">
>       
>       <h5 class="post-reply-notice hidden">
>         <span class="post-reply-arrow"></span><span class="post-reply-name"></span>
>       </h5>
> 
>       
>       <input type="hidden" name="options[entryId]" value="6354f801456983b22c4b91ce42adfca7">
>       <input type="hidden" name="fields[replyThread]" value="">
>       <input type="hidden" name="fields[replyID]" value="">
>       <input type="hidden" name="fields[replyName]" value="">
> 
>       
>       <input required="" name="fields[name]" type="text" placeholder="Your name (Required)" style="background-image: url(&quot;data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAABHklEQVQ4EaVTO26DQBD1ohQWaS2lg9JybZ+AK7hNwx2oIoVf4UPQ0Lj1FdKktevIpel8AKNUkDcWMxpgSaIEaTVv3sx7uztiTdu2s/98DywOw3Dued4Who/M2aIx5lZV1aEsy0+qiwHELyi+Ytl0PQ69SxAxkWIA4RMRTdNsKE59juMcuZd6xIAFeZ6fGCdJ8kY4y7KAuTRNGd7jyEBXsdOPE3a0QGPsniOnnYMO67LgSQN9T41F2QGrQRRFCwyzoIF2qyBuKKbcOgPXdVeY9rMWgNsjf9ccYesJhk3f5dYT1HX9gR0LLQR30TnjkUEcx2uIuS4RnI+aj6sJR0AM8AaumPaM/rRehyWhXqbFAA9kh3/8/NvHxAYGAsZ/il8IalkCLBfNVAAAAABJRU5ErkJggg==&quot;); background-repeat: no-repeat; background-attachment: scroll; background-size: 16px 18px; background-position: 98% 50%; cursor: auto;">
>       <input name="fields[website]" type="text" placeholder="Your website">
>       <input required="" name="fields[email]" type="email" placeholder="Your email address (Required for Gravatar)">
>       <textarea required="" name="fields[body]" placeholder="Your message. Feel free to use Markdown (Google 'Markdown Cheat Sheet')." rows="10"></textarea>
> 
>       
>       
> 
>       
>       <p class="post-submit-notice">
>         <strong class="post-submit-notice-text submit-success hidden">Thanks for your comment! It will show on the site once it has been approved.</strong>
>         <strong class="post-submit-notice-text submit-failed">Sorry, there was an error with your submission.  Please make sure all required fields have been completed and try again.</strong>
>       </p>
> 
>       
>       <input type="submit" value="Submit" class="button">
>       <input type="submit" value="Submitted" class="button hidden" disabled="">
>       <input type="reset" value="Reset" class="button">
>     </form>
> ```

@pacollins To correctly test Staticmn integration for this blog theme,

1. Please post the link to your source repo; https://github.com/pacollins/imperfect-test
2. Please post the output (or at least, a screenshot) from the web browser's web developper tools.  In case of errors, please include the error code and error message in the Staticman API's response, like

    ```js
    {"success":false,"errorCode":"INVALID_VERSION"}
    ```

    This helps us figure out what goes wrong, sometimes.

3. Your repo lacks a [root-level Staticman repo config file](https://staticcman.net/docs/) (see step two in the docs by clicking the link).  That should be copied from https://github.com/pacollins/hugo-future-imperfect-slim/pull/69/files#diff-b7432f3b1e01c9c18f2ba88cb9a82375.
4. You have to place the `staticman.yml` on the *source* branch of your blog instead of the branch holding generated code.  Let's take a closer look to your own Staticman config committed two years ago.
https://github.com/pacollins/hugo-future-imperfect-slim/blob/3e4d5abf5a7c00550611c3e56ac619bbef2443ef/exampleSite/staticman.yml#L42-L45
`data/comments` is a file path on the *source* branch.  On the [branch holding the code generated by `hugo`](https://github.com/pacollins/imperfect-test/tree/7f31edfa9a3a716a0c365c7631087457545d6b12), there's *no* `data/`.

**Comment by pacollins** [comment]
Fixed

> Your repo lacks a root-level Staticman repo config file (see step two in the docs by clicking the link). That should be copied from https://github.com/pacollins/hugo-future-imperfect-slim/pull/69/files#diff-b7432f3b1e01c9c18f2ba88cb9a82375.

>You have to place the staticman.yml on the source branch of your blog instead of the branch holding generated code. Let's take a closer look to your own Staticman config committed two years ago.

We need to move `staticman.yml` to `static/staticman.yml` for builds (or have a second one there.

**Comment by VincentTam** [comment]
> We need to move `staticman.yml` to `static/staticman.yml` for builds (or have a second one there.

I don't get this.  It's clearly said that in the docs that this file has to be root-level.

For a self-contained setup guide, I've recently created a [Wiki page for Staticman setup in Hugo Swift Theme](https://github.com/onweru/hugo-swift-theme/wiki/staticman-config).

I'm sorry that my hands type quite slow.

**Comment by pacollins** [comment]
Well, I take my statement back and clarify it. We need to add documenation somewhere that makes it clear you _must_ put your _entire_ Hugo folder (pre-built and `public`) to a repo. The issue I had was an edge case, but something someone might do. "Hugo builds my site to `public`, so I will just push all of the `public` files to my server." Since `staticman.yml` is at the root of the site, it does _not_ get built into `public`, thus staticman doesn't work. If it were in `static`, it _would_ get pulled into `public` and staticman could still function even if they didn't host the pre-build hugo files.

That all being said, it would cause more issues for them (having to individually download comments), so we (as a theme or as staticman) should probably make sure it is clear you need to push your whole pre-built site to the server and _not_ just public when using staticman.

Does that edge-case thought process make sense? You did nothing wrong, I just represented someone who wasn't thinking and could have misinterpreted the instructions. 

That all being said, I can now happily test your changes. 😄

**Comment by VincentTam** [comment]
> Since `staticman.yml` is at the root of the site, it does _not_ get built into `public`, thus staticman doesn't work. If it were in `static`, it _would_ get pulled into `public` and staticman could still function even if they didn't host the pre-build hugo files.

I wonder what your deploy workflow is.  [My demo site](https://lstu.fr/fish-demo) (see my 2nd comment for the links) is using [GitLab CI](https://framagit.org/VincentTam/fish-demo/pipelines), which gives a publicly accessible file tree of `exampleSite/public`, in which `staticman.yml` is *absent*.  However, my test site can still receive comments.  Why it is so?

Note the difference between "root of the site" and "root of the repository".  In [the Staticman setup](https://staticcman.net/docs/), it shows the later.  The Staticman repo config file `staticman.yml` lies at [the root](https://framagit.org/VincentTam/fish-demo/tree/16beef79b2b74bfbbf1b9d5474f20667377240da) of the *source* branch.

In https://github.com/eduardoboucas/staticman/issues/264#issuecomment-460446511, @<span>eduardobucas</span> wrote

> Staticman doesn't know (and shouldn't care) whether you use GitHub Pages or not. Its job is to get content into your repository. I'm not keen on adding logic that is specific to GitHub Pages.

Staticman is not related to the site.  It communicates with the GitHub/GitLab _repo_.  For a barebone example, you may see https://gitlab.com/ntsim/test-staticman/ (requires GitLab account).  Another theoretically (but not practically) useful example at https://github.com/sdessus/comments_blog_tdm illustrates the functioning of Staticman (as a Node.JS app), even though what the owners claimed in [their blog](https://sdessus.github.io/Commentaires.html) is apparently incorrect.

> Le site est statique, c'est-à-dire que sont contenu est quelque peu figé. Ne vous attendez donc pas à voir apparaître votre commentaire directement après son envoi. Il me faut faire ma petite tambouille pour que vous ayez le bonheur de le voir apparaître sur l'article.
>
> *The site is static.  That means its contents are less movable.  Therefore, don't wait for the appearance of your comment after having sent it.  I have to cook them a little bit so that you will happily see that your comments appears below the article.*

This repo is not practically useful since the comments for all posts have been mixed.  The above claim is wrong since JS's `fetch()` method can send asynchronous requests to remote servers and handle response.

> That all being said, it would cause more issues for them (having to individually download comments), so we (as a theme or as staticman) should probably make sure it is clear you need to push your whole pre-built site to the server and not just public when using staticman.

I prefer using the word "source" over "pre-built".  It has to be, for the moment and near future, on the source branch in a ( custom instance of) GitHub/GitLab repo, since only these two types of remote Git repo are supported.

**Comment by pacollins** [comment]
> I wonder what your deploy workflow is. [My demo site](https://lstu.fr/fish-demo) (see my 2nd comment for the links) is using [GitLab CI](https://framagit.org/VincentTam/fish-demo/pipelines), which gives a publicly accessible file tree of `exampleSite/public`, in which `staticman.yml` is _absent_. However, my test site can still receive comments. Why it is so?

The issue was I _only_ pushed `public` to the repo, thus staticman.yml wasn't included. I personally use Netlify, but for this quick test site, I tried it that way. Again, an edge case. In my case, the root of the repo and site were one in the same. It wasn't a GitHub specific thing, it was how I uploaded (compiled vs source+compiled) - thanks for the better word there by the way. 😉 With staticman.yml in the compiled site folder, I could still take comments, but I would have had to download them individual and place them into my source files. Not practical, but an issue (or intent) someone might have.

> I prefer using the word "source" over "pre-built". It has to be, for the moment and near future, on the source branch in a ( custom instance of) GitHub/GitLab repo, since only these two types of remote Git repo are supported.

Thanks for the word I couldn't think of when typing. 😄 Yes, it should be clear that it should exist within the source branch (not just the compiled site) on a repo.

**Comment by VincentTam** [comment]
Notes for commits after @<span>pacollins</span>'s approval:

The `clearForm()` function had a bug: after successful form submission, the visible form fields weren't reset.
https://github.com/pacollins/hugo-future-imperfect-slim/blob/5e9723377b352c7280975c00a3c8e8da0eaec9e4/assets/js/staticman.js#L23-L27
This is due to the extra `[type="hidden"]` at line 52.
https://github.com/pacollins/hugo-future-imperfect-slim/blob/5e9723377b352c7280975c00a3c8e8da0eaec9e4/assets/js/staticman.js#L50-L59
I had made such mistake after hours of programming, and I had forgotten that `clearForm()` is also called in caes of successful form submission.  I had the wrong idea that the `onclick` listener of the "Reset" button were the only function that called `clearForm()`.
https://github.com/pacollins/hugo-future-imperfect-slim/blob/5e9723377b352c7280975c00a3c8e8da0eaec9e4/assets/js/staticman.js#L100-L103

Now this has been corrected at ffa6711.
https://github.com/pacollins/hugo-future-imperfect-slim/blob/ffa671149fdf9fd008a4e56237a594ae7a6f560c/assets/js/staticman.js#L50-L60

In addition, since a Staticman instance running on a free Heroku dyno (say, @staticmanlab) might take time to respond, I've added a SCSS rule to decrease the opacity of the submitted form by half until the submission message appears.  I've changed `$(form).disable = true` to `$(form).addClass('loading')` because the [`disabled` attribute doesn't apply to `<form>` tags](https://www.w3schools.com/TAGS/att_disabled.asp).
